ambry-store/src/main/java/com.github.ambry.store/Log.java

```diff
@@ -13,161 +13,304 @@
-    fileChannel.close();
+    // TODO (DiskManager changes): This is intended to "request" the segment file from the DiskManager which will have
+    // TODO (DiskManager changes): a pool of segments.
+    File segmentFile = new File(dataDir, filename);
+    Utils.preAllocateFileIfNeeded(segmentFile, size);
+    remainingUnallocatedCapacity.addAndGet(-size);
+    return segmentFile;
   }
 
-  public void flush()
+  /**
+   * Rolls the log segment over if required. Also ensures enough capacity.
+   * @param writeSize the size of the incoming write.
+   * @throws IllegalArgumentException if the {@code writeSize} is greater than a single segment's size
+   * @throws IllegalStateException if there no more capacity in the log.
+   * @throws IOException if any I/O error occurred as a part of ensuring capacity.
+   *
+   */
+  private void rollOverIfRequired(long writeSize)
       throws IOException {
-    fileChannel.force(true);
+    if (activeSegment.getCapacityInBytes() - activeSegment.getEndOffset() < writeSize) {
+      ensureCapacity(writeSize);
+      // this cannot be null since capacity has either been ensured or has thrown.
+      LogSegment nextActiveSegment = segmentsByName.higherEntry(activeSegment.getName()).getValue();
+      logger.info("Rolling over writes to {} from {} on write of data of size {}. End offset was {} and capacity is {}",
+          nextActiveSegment.getName(), activeSegment.getName(), writeSize, activeSegment.getEndOffset(),
+          activeSegment.getCapacityInBytes());
+      activeSegment = nextActiveSegment;
+    }
   }
 
-  @Override
-  public void readInto(ByteBuffer buffer, long position)
+  /**
+   * Ensures that there is enough capacity of a write of size {@code writeSize}.
+   * @param writeSize the size of a subsequent write on the active log segment.
+   * @throws IllegalArgumentException if the {@code writeSize} is greater than a single segment's size
+   * @throws IllegalStateException if there no more capacity in the log.
+   * @throws IOException if any I/O error occurred as a part of ensuring capacity.
+   */
+  private void ensureCapacity(long writeSize)
       throws IOException {
-    if (sizeInBytes() < position || (position + buffer.remaining() > sizeInBytes())) {
-      metrics.overflowReadError.inc(1);
-      logger.error("Log: {} Error trying to read outside the log range. log end position {} input buffer size {}",
-          file.getAbsolutePath(), sizeInBytes(), buffer.remaining());
-      throw new IllegalArgumentException("Log : " + file.getAbsolutePath() + " error trying to read outside " +
-          "the log range. log end position " + sizeInBytes() + " input buffer size " + buffer.remaining());
+    if (segmentsByName.higherEntry(activeSegment.getName()) != null) {
+      return;
+    }
+    // all segments are (should be) the same size.
+    long segmentCapacity = activeSegment.getCapacityInBytes();
+    if (segmentCapacity > remainingUnallocatedCapacity.get()) {
```

Comment:
this is not very intuitive. far more intuitive to check for `remainingUnallocatedCapacity.get()  == 0`. Isn't that really what you are trying to do here? Might be even simpler to just use `remainingUnallocatedSegments` as the value only gets decremented in multiples of `segmentCapacity`.